testutils/check-lang-uses-form/main.go

```diff
@@ -0,0 +1,185 @@
+`
+
+func main() {
+	formFname := flag.String("forms", "", "file containing calling forms")
+	flag.Usage = func() {
+		binName := os.Args[0]
+		out := os.Stderr // Should be flag.CommandLine.Output(), but Go version on Travis is too old.
+		fmt.Fprintf(out, "Usage of %s:\n", binName)
+		fmt.Fprintf(out, "%s -form CallingForm.java [dir]\n", binName)
+		flag.PrintDefaults()
+		fmt.Fprint(out, usage)
+	}
+	flag.Parse()
+
+	if *formFname == "" {
+		flag.Usage()
+		os.Exit(1)
+	}
+
+	formf, err := os.Open(*formFname)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer formf.Close()
+
+	checks, err := readChecks(formf)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	walkFn := func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if len(checks) == 0 {
+			return filepath.SkipDir
+		}
+		if !info.Mode().IsRegular() || filepath.Ext(path) != ".baseline" {
+			return nil
+		}
+
+		lang := filepath.Base(filepath.Dir(path))
+		words := checks[lang]
+		if len(words) == 0 {
+			return nil
+		}
+
+		baseline, err := ioutil.ReadFile(path)
+		if err != nil {
+			return err
+		}
+
+		words = notFoundWords(string(baseline), words)
+		if len(words) == 0 {
+			delete(checks, lang)
+		} else {
+			checks[lang] = words
+		}
+		return nil
+	}
+
+	dir := "."
+	if flag.NArg() > 0 {
+		dir = flag.Arg(0)
+	}
+	if err := filepath.Walk(dir, walkFn); err != nil {
+		log.Fatal(err)
+	}
+
+	if len(checks) != 0 {
+		fmt.Println("The following {language, callingForms}s are declared but not used:")
+		for lang, forms := range checks {
+			for _, form := range forms {
+				fmt.Printf("{%q, %q}\n", lang, form)
+			}
+		}
+		os.Exit(1)
+	}
+}
+
+// readChecks reads calling forms and languages where calling forms are expected in from r.
+// The format is described by the program's usage text.
+func readChecks(r io.Reader) (map[string][]string, error) {
+	checks := make(map[string][]string)
+	sc := bufio.NewScanner(r)
+	for sc.Scan() {
+		ln := sc.Text()
+		p := strings.Index(ln, "//")
+		const prefix = "// used by:"
+		if p < 0 || !strings.HasPrefix(ln[p:], prefix) {
+			continue
+		}
+		form := ln[:p]
+		form = strings.TrimSpace(form)
+		form = strings.TrimRight(form, ",")
+
+		for _, lang := range strings.Fields(ln[p+len(prefix):]) {
+			checks[lang] = append(checks[lang], form)
+		}
+	}
+	if err := sc.Err(); err != nil {
+		return nil, err
+	}
+	return checks, nil
+}
+
+// notFoundWords searches s for words and report words not found in s.
+// A word must be delimited by either string boundary or non-word character,
+// eg string "ab" does not contain the word "b".
+func notFoundWords(s string, words []string) []string {
+	var notFound []string
+	for _, w := range words {
+		if !hasWord(s, w) {
+			notFound = append(notFound, w)
+		}
+	}
+	return notFound
+}
+
+// hasWord reports wheter s contains word w. See notFoundWords for details.
+func hasWord(s, w string) bool {
+	p := 0
```

Comment:
Why not implement this via a regex, which will be easier to follow and verify? For the word boundaries, I think you could even use one of the predefined character classes in https://github.com/google/re2/wiki/Syntax#ascii, since those are supported by https://golang.org/pkg/regexp/ Ex:
```
pattern := fmt.Sprintf("[^[[:word:]]]+%s[^[[:word:]]]+",w)
```

Reply:
I thought about this and decided against it. Philosophically, regexps are useful for matching patterns known at runtime. When the pattern, or at least the format, is simple and known ahead of time, I'd argue that using the regexp is a bit of an overkill.

Technically (though it doesn't really matter here), regexp is usually slow even when implemented properly:
```
BenchmarkRegex-12      	   10000	    148171 ns/op	  81.06 MB/s
BenchmarkHasWord-12    	10000000	       147 ns/op	81468.69 MB/s

```

Reply:
This is not performance-critical, and the regex is still sub millisecond, which is fine for what we're doing. In that case, the important criterion becomes clarity. People know regexs and they're easy to read (much more compact) and think about, as opposed to double-checking carefully this custom method.

Something to think about. Not going to block on this.